docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,369 @@
+If operator does not wish to wait for the scheduled full/delta snapshot, he/she can trigger an on-demand(out-of-schedule) full/delta snapshot on etcd cluster which will be taken by `leading-backup-restore`.
+
+##### Possible scenarios
+
+* [Shoot hibernation](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_hibernate.md): Every etcd cluster incurs an inherent cost of preserving the volumes even when the shoot is in hibernation state. However it is possible to save costs by invoking this task to take a full snapshot before deleting the volumes.
+* [Control Plane Migration](https://github.com/gardener/gardener/blob/master/docs/proposals/07-shoot-control-plane-migration.md): In [gardener](https://github.com/gardener/gardener) a seed cluster control plane can be moved to another seed cluster. To prevent data loss and faster restoration of the etcd cluster in the target seed, a full snapshot can be triggered for the etcd cluster in the source seed.
+* A on-demand full snapshot can be triggered if scheduled snapshot fails due to any reason.
+
+##### Task Config
+
+```go
+// SnapshotType can be full or delta snapshot
+SnapshotType string `json:"snapshotType"`
+const (
+  FullSnapshot SnapshotType = "full-snapshot"
+  DeltaSnapshot SnapshotType = "delta-snapshot"
+)
+
+type OnDemandSnapshotTaskConfig struct {
+  // Type of on-demand snapshot.
+  SnapshotType snapshotType `json:"snapshotType"`
+}
+```
+
+##### Pre-Conditions
+
+* Etcd cluster should have a quorum.
+* There should not already be a `on-demand snapshot` task running with the same `SnapshotType` for the same etcd cluster.
+
+#### Trigger on-demand maintenance of etcd cluster
+
+Operator can trigger on-demand maintenance of etcd cluster which includes operations like [etcd compaction](https://etcd.io/docs/v3.5/op-guide/maintenance/#history-compaction-v3-api-key-value-database), [etcd defragmentation](https://etcd.io/docs/v3.2/op-guide/maintenance/#defragmentation) etc.
+
+##### Possible Scenarios
+
+* If an etcd cluster is heavily loaded which is causing performance degradation of an etcd cluster and the operator does not want to wait for the scheduled maintenance window then an on-demand maintenance task can be triggered which will invoke etcd-compaction, etcd-defragmentation etc. on the target etcd cluster. This will make etcd cluster lean and clean, thus improving cluster performance.
+
+##### Task Config
+
+```go
+type OnDemandMaintenanceTaskConfig struct {
+  // MaintenanceType defines the maintenance operations need to be performed on etcd cluster.
+  MaintenanceType maintenanceOps
+}
+
+type maintenanceOps struct {
+  EtcdCompaction *bool `json:"etcd-compaction,omitempty"`
```

Comment:
Even though the fields are self explanatory, for consistency's sake it would be nicer to have a comment describing each field, like there is for every other struct definition.